src/client/consumer.rs

```diff
@@ -186,38 +232,74 @@ impl Stream for StreamConsumer {
             }
 
             if this.fetch_fut.is_terminated() {
-                let offset = *this.next_offset;
+                let next_offset = *this.next_offset;
+                let start_offset = *this.start_offset;
                 let bytes = (*this.min_batch_size)..(*this.max_batch_size);
                 let max_wait_ms = *this.max_wait_ms;
                 let client = Arc::clone(this.client);
 
-                trace!(offset, "Fetching records at offset");
+                trace!(?start_offset, ?next_offset, "Fetching records at offset");
 
                 *this.fetch_fut = FutureExt::fuse(Box::pin(async move {
-                    client.fetch_records(offset, bytes, max_wait_ms).await
+                    let offset = match next_offset {
+                        Some(x) => x,
+                        None => match start_offset {
+                            StartOffset::Earliest => client.get_offset(OffsetAt::Earliest).await?,
+                            StartOffset::Latest => dbg!(client.get_offset(OffsetAt::Latest).await?),
+                            StartOffset::At(x) => x,
+                        },
+                    };
+
+                    let (records_and_offsets, watermark) =
+                        client.fetch_records(offset, bytes, max_wait_ms).await?;
+                    Ok(FetchResultInner {
+                        records_and_offsets,
+                        watermark,
+                        used_offset: offset,
+                    })
                 }));
             }
 
             let data: FetchResult = futures::ready!(this.fetch_fut.poll_unpin(cx));
 
-            match data {
-                Ok((mut v, watermark)) => {
+            match (data, *this.start_offset) {
+                (Ok(inner), _) => {
+                    let FetchResultInner {
+                        mut records_and_offsets,
+                        watermark,
+                        used_offset,
+                    } = inner;
                     trace!(
                         high_watermark = watermark,
-                        n_records = v.len(),
+                        n_records = records_and_offsets.len(),
                         "Received records and a high watermark",
                     );
 
+                    // Remember used offset (might be overwritten if there was any data) so we don't refetch the
+                    // earliest / latest offset for every try. Also fetching the latest offset might be racy otherwise,
+                    // since we'll never be in a position where the latest one can actually be fetched.
+                    *this.next_offset = Some(used_offset);
```

Comment:
So the issue here is that "latest" offset: If you have a new or empty partition, then Kafka returns the start of that empty interval (e.g. `0` if it's an empty partition) if you ask it, also see

https://github.com/influxdata/rskafka/blob/15f424f20f9da982ce79bf8fd74e4d32cfde876c/tests/client.rs#L241-L251

Now if you don't remember that offset, the time your produce something to the partition the "latest" offset (which is the latest record + 1) will also increase, so you're always chasing the tail of the partition but never fetch any records. So when you first get an answer to "what's the earliest/latest offset?" you need to remember that (except when this turns out to be wrong due to "offset out of range").

When the partition stays empty, then sure this loops forever and asks Kafka again and again for data. However this is not a hot loop since the fetch request has a timeout attached, so the broker will wait some milliseconds until it tells us "there are no records here, like the last X time you've asked me".